project_euler/problem_085/sol1.py

```diff
@@ -0,0 +1,102 @@
+    triangle numbers. So to find the solution grid (a,b), we need to find the two
+    triangle numbers whose product is closest to two million.
+
+    Denote these two triangle numbers Ta and Tb. We want their product Ta*Tb to be
+    as close as possible to 2m. Assuming that the best solution is fairly close to 2m,
+    We can assume that both Ta and Tb are roughly bounded by 2m. Since Ta = a(a+1)/2,
+    we can assume that a (and similarly b) are roughly bounded by sqrt(2 * 2m) = 2000.
+    Since this is a rough bound, to be on the safe side we add 10%. Therefore we start
+    by generating all the triangle numbers Ta for 1 <= a <= 2200. This can be done
+    iteratively since the ith triangle number is the sum of 1,2, ... ,i, and so
+    T(i) = T(i-1) + i.
+
+    We then search this list of triangle numbers for the two that give a product
+    closest to our target of two million. Rather than testing every combination of 2
+    elements of the list, which would find the result in quadratic time, we can find
+    the best pair in linear time.
+
+    We iterate through the list of triangle numbers using enumerate() so we have a
+    and Ta. Since we want Ta * Tb to be as close as possible to 2m, we know that Tb
+    needs to be roughly 2m / Ta. Using the formula Tb = b*(b+1)/2 as well as the
+    quadratic formula, we can solve for b:
+    b is roughly (-1 + sqrt(1 + 8 * 2m / Ta)) / 2.
+
+    Since the closest integers to this estimate will give product closest to 2m,
+    we only need to consider the integers above and below. It's then a simple matter
+    to get the triangle numbers corresponding to those integers, calculate the product
+    Ta * Tb, compare that product to our target 2m, and keep track of the (a,b) pair
+    that comes the closest.
+
+
+Reference: https://en.wikipedia.org/wiki/Triangular_number
+           https://en.wikipedia.org/wiki/Quadratic_formula
+"""
+
+
+from math import ceil, floor, sqrt
+from typing import List
+
+
+def solution(target: int = 2000000) -> int:
+    """
+    Find the area of the grid which contains as close to two million rectangles
+    as possible.
+    >>> solution(20)
+    6
+    >>> solution(2000)
+    72
+    >>> solution(2000000000)
+    86595
+    """
+    triangle_numbers: List[int] = [0]
+    idx: int
+
+    for idx in range(1, ceil(sqrt(target * 2) * 1.1)):
+        triangle_numbers.append(triangle_numbers[-1] + idx)
+
+    best_product: int = 0  # we want this to be as close as possible to target
+    area: int = 0  # the area a*b corresponding to the grid that gives
+    # the product closest to target
+    b_estimate: float  # an estimate of b, using the quadratic formula
+    b_floor: int  # the largest integer less than b_estimate
+    b_ceil: int  # the largest integer less than b_estimate
+    triangle_b_first_guess: int  # the triangle number corresponding to b_floor
+    triangle_b_second_guess: int  # the triangle number corresponding to b_ceil
```

Comment:
One line comments should always be on the top. If there are only a few words, then they can be on the same line. This is done to provide better readability :)

Reply:
Is that better?

Reply:
Oh, I am sorry to be ambiguous but I meant on top of the variable itself instead of on the side.